File: writing-core/src/main/java/dev/bee/kanjianki/study/WritingRecognizer.java
package dev.bee.kanjianki.study;

import dev.bee.kanjianki.core.study.RecognitionCandidate;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public interface WritingRecognizer extends AutoCloseable {
    CompletableFuture<ModelStatus> modelStatus();

    CompletableFuture<ModelStatus> downloadModel();

    CompletableFuture<RecognitionResult> recognize(CapturedWriting writing);

    @Override
    void close();

    static List<RecognitionCandidate> recognitionCandidates(RecognitionResult result) {
        return result == null ? Collections.emptyList() : result.toRecognitionCandidates();
    }

    final class ModelStatus {
        public final String modelName;
        public final String languageTag;
        public final boolean downloaded;
        public final String message;

        public ModelStatus(String modelName, String languageTag, boolean downloaded, String message) {
            this.modelName = modelName;
            this.languageTag = languageTag;
            this.downloaded = downloaded;
            this.message = message;
        }
    }

    final class RecognitionResult {
        public final List<Candidate> candidates;

        public RecognitionResult(List<Candidate> candidates) {
            this.candidates = Collections.unmodifiableList(new ArrayList<>(candidates));
        }

        public String topText() {
            if (candidates.isEmpty()) {
                return "";
            }
            return candidates.get(0).text;
        }

        public List<RecognitionCandidate> toRecognitionCandidates() {
            List<RecognitionCandidate> out = new ArrayList<>();
            for (Candidate candidate : candidates) {
                out.add(new RecognitionCandidate(candidate.text, candidate.score));
            }
            return out;
        }
    }

    final class Candidate {
        public final String text;
        public final Float score;

        public Candidate(String text, Float score) {
            this.text = text == null ? "" : text;
            this.score = score;
        }
    }
}

File: writing-core/src/main/kotlin/dev/bee/kanjianki/study/WritingRecognizer.kt
package dev.bee.kanjianki.study

import dev.bee.kanjianki.core.study.RecognitionCandidate
import java.util.ArrayList
import java.util.Collections
import java.util.concurrent.CompletableFuture

interface WritingRecognizer : AutoCloseable {
    fun modelStatus(): CompletableFuture<ModelStatus>

    fun downloadModel(): CompletableFuture<ModelStatus>

    fun recognize(writing: CapturedWriting?): CompletableFuture<RecognitionResult>

    override fun close()

    class ModelStatus(
        @JvmField val modelName: String?,
        @JvmField val languageTag: String?,
        @JvmField val downloaded: Boolean,
        @JvmField val message: String?,
    )

    class RecognitionResult(candidates: List<Candidate>) {
        @JvmField
        val candidates: List<Candidate> = Collections.unmodifiableList(ArrayList(candidates))

        fun topText(): String {
            if (candidates.isEmpty()) {
                return ""
            }
            return candidates[0].text
        }

        fun toRecognitionCandidates(): List<RecognitionCandidate> {
            val out = ArrayList<RecognitionCandidate>()
            for (candidate in candidates) {
                out.add(RecognitionCandidate(candidate.text, candidate.score))
            }
            return out
        }
    }

    class Candidate(text: String?, @JvmField val score: Float?) {
        @JvmField
        val text: String = text ?: ""
    }

    companion object {
        @JvmStatic
        fun recognitionCandidates(result: RecognitionResult?): List<RecognitionCandidate> {
            return result?.toRecognitionCandidates() ?: Collections.emptyList()
        }
    }
}
